### ISSUE TITLE
keepdims option for sum, std, sem, var, mean etc.

### ISSUE THREAD

**Issue by tachukao** [owlbarn/owl#529]
I often find myself writing code such as 
```
let mu = Arr.mean ~axis:0 x |> Arr.squeeze ~axis:[|axis|]
let stdev = Arr.std ~axis:0 x |> Arr.squeeze ~axis:[|0|]
```
It will be useful to have a `keepdims` option as in numpy [`sum`](https://numpy.org/doc/stable/reference/generated/numpy.sum.html), which automatically squeezes out the dimension we `sum` over if `keepdims` is set `true`.

In numpy, `keepdims` is true by default. For backward compatibility, it is probably best to make `false` the default of `keepdims` in Owl if we make these changes. Such changes should be straightforward to incorporate into Owl's Ndarray and Algodiff modules. After the proposed changes we will be able to write
```
let mu = Arr.mean ~axis:0 ~keepdims:false x
```

@ryanrhymes  @mseri  @ghennequin  @jzstark  thoughts?

**Comment by mseri** [comment]
This sounds like a welcome change to me.

At this stage I think we should not be too worried about backward compatibility, and think more about having an ergonomic and consistent API before we draft a stable release.

I personally think `keepdims:true` is more convenient, but we need to test the repercussions of such change very carefully.

**Comment by ghennequin** [comment]
a minor thing: `keepdims` should probably be `keep_dims` for general consistency with how distinct words are separated

**Comment by ryanrhymes** [comment]
I agree, it looks like a very good addition to current function parameters. No need to be backward compatibility as Marcello said. :)

**Comment by tachukao** [comment]
In my original post I completely messed up what I wanted to say.
The default of `keepdims` is `false` in numpy  (I said it's true in the original post) and for backward compatibility, it's best to have `keep_dims` as `true` in Owl. 
We squeeze the dimension if `keep_dims` is set to false (I said we squeeze it out if set to true in the original post).
I agree with @mseri that `keep_dims:true` is probably more convenient as will be backward compatible (unless that's not what you meant to say @mseri ?)

**Comment by mseri** [comment]
I meant that backward compatibility is not a priority and I would not worry about that.

I find the most convenient default is the one that squeezes out the dimensions. Now I am not sure which one of them it is

**Comment by tachukao** [comment]
@mseri argh okay so I guess we disagree then, I find the _not_ squeezing out the dimension a more convenient default, and is how I've coded it up in the new PR #531. But I also don't really have a strong opinion about this, happy to change it back if others feel the same @ryanrhymes @ghennequin ?

**Comment by bluddy** [comment]
If there isn't a clear consensus or preference, it probably makes sense to go by whatever numpy does by default, so people's expectations are met.

**Comment by ghennequin** [comment]
As @tachukao, I find that _not_ squeezing is my most frequent use case, but I'm not too fussed either.

**Comment by ryanrhymes** [comment]
I am really neutral on this.
